### ISSUE TITLE
Question Regarding Refiner Network

### ISSUE THREAD

**Issue by HONGcogml** [alcros33/ThermalSuperResolution#1]
I have read your paper on applying Diffusion Models to TISR with great interest, and it has been very helpful for my own research.

I tested your model using the CIDIS 2025 track-2 x8 validation dataset. The pretrained weights I used were downloaded from your repository:

[https://github.com/alcros33/ThermalSuperResolution/releases/tag/weights](url)
- resshift_thermal_track2_x8.ckpt
- unet_refiner_track2_x8.ckpt

The inference with the **ResShift model** produced correct results. The command I used was:

> python inference.py resshift_thermal_track2_x8.ckpt --cuda --bs 8 \
>     -i CIDIS/thr/val/LR_x8 CIDIS/rgb/val -o results/0912-ResShift

However, when I performed the final inference with the **Refiner model**, the reconstructed images did not preserve the thermal image characteristics. Instead, they appeared with incorrect colorization, and both PSNR and SSIM scores were lower compared to the ResShift results. The command I used was:

> python inference.py ./unet_refiner_track2_x8.ckpt --refiner --cuda --bs 8 \
>     -i CIDIS/thr/val/LR_x8 CIDIS/rgb/val results/0912-ResShift -o results/0912-Refiner

Please note that I did not modify any of the provided configs YAML files or Python scripts in the repository.

I have attached example images for reference.

> [ResShift-001_01_D4.bmp](https://github.com/user-attachments/files/22291985/ResShift-001_01_D4.bmp)
> [Refiner-001_01_D4.bmp](https://github.com/user-attachments/files/22291989/Refiner-001_01_D4.bmp)
> 
> [ResShift-019_01_D3.bmp](https://github.com/user-attachments/files/22291991/ResShift-019_01_D3.bmp)
> [Refiner-019_01_D3.bmp](https://github.com/user-attachments/files/22291993/Refiner-019_01_D3.bmp)

Could you kindly advise on what might be causing this issue?

Thank you very much for your time and guidance.

Best regards,
ChaeHui Hong

**Comment by alcros33** [comment]
I think I might have forgot to add a line to correctly save the image. I don't really remember why we implemented it like this but if you look into unet_refiner_track1.yaml you can see that output_channels is 3. So images are predicted in rgb but we averaged the channels at the end to form a 1 channel result. I think it had better numeric stability. So you may fix this by adding one line after line 51 in inference.py `pred = pred.mean(dim=1, keepdim=True)` if you get another error when saving the image you may need to repeat the one channel 3 times (it is dumb) `pred = pred.repeat(1,3,1,1)`

**Comment by HONGcogml** [comment]
Hello,

First of all, I’m pleased to provide feedback on the Refiner model from your repository, which I used for my experiments.

I conducted experiments using the code you provided, and the average PSNR of the Refiner model's reconstructed image was 17.35, with an SSIM of 0.4799. After applying the channel averaging method you suggested to the _inference.py_ code, the PSNR and SSIM values decreased instead.

To identify the problem, I carefully analyzed the repository code and discovered an error in the inference command written in the README. Specifically, in the _Unet_refiner.py_, the input for the _forward_ function should be **_torch.cat([up_low_res, rough_pred, visible], dim=1)_**, but in the command, the order of arguments for _**LR, visible,**_ and _**rough_results**_ was reversed.

After correcting the command as follows, I was able to obtain the correct Refiner reconstruction images (ex, [correct_refiner_019_01_D3.bmp](https://github.com/user-attachments/files/22309926/correct_refiner_019_01_D3.bmp)) :

> python inference.py refiner_checkpoint.pth --refiner --bs 8 -i LR rough_results visible -o results

Additionally, I compared the Refiner experimental results, and without averaging, the average PSNR was 27.03 with SSIM 0.8105. After applying the averaging, the average PSNR slightly increased to 27.04 with SSIM 0.8116.

In conclusion, the issue with the colorization of the Refiner model was due to the reversed order of **_visible_** and **_rough_results_** in the inference command from the repository. I’m glad I was able to resolve the issue.

Additionally, I would like to **respectfully inquire** if it would be **permissible** to use the **Refiner network** from your research in my own study. Your excellent research materials have been incredibly helpful to my work, and I would like to express my **sincere gratitude**.

Thank you,
ChaeHui Hong

**Comment by alcros33** [comment]
Thank you for using my work on your research, by all means feel free to use my code in your work. Just please be sure to add the paper in your citations
